velox/functions/prestosql/ArithmeticImpl.h

```diff
@@ -132,4 +132,14 @@ T ceil(const T& arg) {
   return results;
 }
 
+template <typename TNum, typename TDecimals>
+FOLLY_ALWAYS_INLINE TNum
+truncate(const TNum& number, const TDecimals& decimals = 0) {
+  if (std::isnan(number) || std::isinf(number) || decimals == 0) {
+    return std::trunc(number);
+  }
+  const auto factor = std::pow(10, decimals);
```

Comment:
is this how this is implemented in the java codebase?

Reply:
Not exactly. The java implementation relies on BigDecimal(setScale, valueOf etc.) and not very simple logic incapsulated in it. To my research we don't possess similar functionality in c++.  Unless we want to implemet/grab BigDecimal like utilities from somewhere this seems to be the quickest way of having the function available(though with the risk of mismatches in some cases).

Side question, do we have a A/B engine tests setup to test between presto and velox operators?

Reply:
We support Decimal types in Velox (both Big and Small). The functionality should be similar but the API might not be exactly the same. 

> Side question, do we have a A/B engine tests setup to test between presto and velox operators?

There's a test suite in Prestissimo that validates that C++ and Java versions return the same results, but the queries need to be manually written there. We should make sure we add something covering truncate behavior as well.

Reply:
So we are interested in mimicking this line https://github.com/prestodb/presto/blob/ba9ec306d453f968f711e437ab2d60b91dc6b700/presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java#L380. Specifically:

1. **valueOf()** - https://github.com/openjdk-mirror/jdk7u-jdk/blob/f4d80957e89a19a29bb9f9807d2a28351ed7f7df/src/share/classes/java/math/BigDecimal.java#L806
2. **setScale()** - https://github.com/openjdk-mirror/jdk7u-jdk/blob/f4d80957e89a19a29bb9f9807d2a28351ed7f7df/src/share/classes/java/math/BigDecimal.java#L2378
3. **doubleValue()** - https://github.com/openjdk-mirror/jdk7u-jdk/blob/f4d80957e89a19a29bb9f9807d2a28351ed7f7df/src/share/classes/java/math/BigDecimal.java#L3132
4. And **toString** as an implementation detail

We have:
1. **UnscaledLongDecimal** - https://github.com/facebookincubator/velox/blob/a19a652939f98e8a9cfd788d15976a8edd999c62/velox/type/UnscaledLongDecimal.h#L55
2. **UnscaledShortDecimal** - https://github.com/facebookincubator/velox/blob/a19a652939f98e8a9cfd788d15976a8edd999c62/velox/type/UnscaledShortDecimal.h#L26
3. **DecimalUtils** with rescale and toString - https://github.com/facebookincubator/velox/blob/a19a652939f98e8a9cfd788d15976a8edd999c62/velox/type/DecimalUtil.h#L28

Looking into how to leverage those to mimick Java behavior.